addons/adyen_marketpay/models/adyen_account.py

```diff
@@ -0,0 +1,640 @@
+            })
+            checks = data.get('verification', {})
+            all_checks_status = []
+
+            # Account Holder Checks
+            account_holder_checks = checks.get('accountHolder', {})
+            account_holder_errors = []
+            for check in account_holder_checks.get('checks'):
+                all_checks_status.append(check['status'])
+                if check['status'] not in ['PASSED', 'DATA_PROVIDED', 'PENDING']:
+                    account_holder_errors.append(self._get_kyc_message(check))
+
+            # Shareholders Checks
+            shareholder_checks = checks.get('shareholders', {})
+            shareholder_errors = []
+            kyc_error_message = False
+            for sc in shareholder_checks:
+                shareholder_status = []
+                shareholder_id = adyen_account_id.shareholder_ids.filtered(lambda shareholder: shareholder.shareholder_uuid == sc['shareholderCode'])
+                for check in sc.get('checks'):
+                    all_checks_status.append(check['status'])
+                    shareholder_status.append(check['status'])
+                    if check['status'] not in ['PASSED', 'DATA_PROVIDED', 'PENDING']:
+                        kyc_error_message = self._get_kyc_message(check)
+                        shareholder_errors.append('[%s] %s' % (shareholder_id.display_name, kyc_error_message))
+                shareholder_id.with_context(update_from_adyen=True).write({
+                    'kyc_status': self.get_status(shareholder_status),
+                    'kyc_error_message': kyc_error_message,
+                })
+
+            # Bank Account Checks
+            bank_account_checks = checks.get('bankAccounts', {})
+            bank_account_errors = []
+            kyc_error_message = False
+            for bac in bank_account_checks:
+                bank_account_status = []
+                bank_account_id = adyen_account_id.bank_account_ids.filtered(lambda bank_account: bank_account.bank_account_uuid == bac['bankAccountUUID'])
+                for check in bac.get('checks'):
+                    all_checks_status.append(check['status'])
+                    bank_account_status.append(check['status'])
+                    if check['status'] not in ['PASSED', 'DATA_PROVIDED', 'PENDING']:
+                        kyc_error_message = self._get_kyc_message(check)
+                        bank_account_errors.append('[%s] %s' % (bank_account_id.display_name, kyc_error_message))
+                bank_account_id.with_context(update_from_adyen=True).write({
+                    'kyc_status': self.get_status(bank_account_status),
+                    'kyc_error_message': kyc_error_message,
+                })
+
+            kyc_status = self.get_status(all_checks_status)
+            if kyc_status in ['failed', 'awaiting_data']:
+                kyc_error_message = self.env['ir.qweb'].render('adyen_marketpay.kyc_error_message', {
```

Comment:
Don't you think we could create a next activity on the account adyen model, when there is a kyc error. (only do it once, not each time the cron runs)

Reply:
Yes, otherwise nobody will be notified that some data is requested. I talked about this with Laura last week and she suggested that we log messages instead of creating an activity, this way the user will still be notified since he's a follower of the account but we can also keep track of all the validation steps. I don't know what's the best option?

Reply:
Ok for the message log, it is a good idea, but we must be sure to log one only if the kyc_status and/or message changes to avoid writing useless message.

Maybe we should write messages only on the acount object, even if it is related to a shareholder or bank account to keep account status in only one place?